src/components/header_group/header_group_container.ts

```diff
@@ -0,0 +1,116 @@
+import { Component, useState } from "@odoo/owl";
+import { Action } from "../../actions/action";
+import { FROZEN_PANE_HEADER_BORDER_COLOR, GROUP_LAYER_WIDTH } from "../../constants";
+import { createHeaderGroupContainerContextMenu } from "../../registries/menus/header_group_registry";
+import { DOMCoordinates, SpreadsheetChildEnv } from "../../types";
+import { CSSProperties, Dimension, HeaderGroup, Pixel } from "../../types/misc";
+import { css, cssPropertiesToCss } from "../helpers";
+import { HEADER_HEIGHT, HEADER_WIDTH } from "./../../constants";
+import { Menu } from "./../menu/menu";
+import { ColGroup, RowGroup } from "./header_group";
+
+interface Props {
+  dimension: Dimension;
+  layers: HeaderGroup[][];
+}
+
+interface State {
+  menuState?: { menuItems: Action[]; position: DOMCoordinates | undefined };
+}
+
+css/* scss */ `
+  .o-header-group-frozen-pane-border {
+    &.o-group-rows {
+      margin-top: -1px;
+      border-bottom: 3px solid ${FROZEN_PANE_HEADER_BORDER_COLOR};
+    }
+    &.o-group-columns {
+      margin-left: -1px;
+      border-right: 3px solid ${FROZEN_PANE_HEADER_BORDER_COLOR};
+    }
+  }
+
+  .o-header-group-main-pane {
+    &.o-group-rows {
+      margin-top: -2px; // Counteract o-header-group-frozen-pane-border offset
+    }
+    &.o-group-columns {
+      margin-left: -2px;
+    }
+  }
+`;
+
+export class HeaderGroupContainer extends Component<Props, SpreadsheetChildEnv> {
+  static template = "o-spreadsheet-HeaderGroupContainer";
+  static components = { RowGroup, ColGroup, Menu };
+
+  state: State = useState({ menuState: undefined });
```

Comment:
why `this.state.menuState` instead of directly `this.menuState` ?

Reply:
Having `this.menuState` means that I need to have a boolean `isMenuActive` in the `menuState` to toggle the menu, rather than having `this.state.menuState` be undefined.

But I agree that it's ugly. Not sure which I prefer. I can change if you want.

Reply:
I would go with an `isMenuActive`. It also better from semantic POV to have a real boolean rather than comparing with undefined.
Btw, `menuState` could even be named `menu`. `this.menu.isActive`, `this.menu.position`, etc. is looking nice